### PR TITLE
Introduced "Blocks" to fix the bug #1760

### PR DESCRIPTION
**Pull request by guy-maurel**
To align (or not align) the assign character it is important to know if:
    1. the levels of the assign chunks are the same
    2. the block numbers of the statements are the same.
 
Introducing block numbering as:
 A new block is opened if the type of the chunk is:
```.txt
    CT_BRACE_OPEN
    CT_FPAREN_OPEN
    CT_ANGLE_OPEN
```

 With this we get:
```.cpp
    virtual void f(int x, int y) = 133;
    void g(int x = 144);
```

 and not (Issue #1760)
```.cpp
    virtual void f(int x, int y) = 133;
    void g(int x                 = 144);

```
Add some more LOG_FMT statements.
Update the result of test cpp/30023

### CODE REVIEW

File: src/BlockNumbering.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+#include "BlockNumbering.h"
+#include "chunk_list.h"
+#include "log_levels.h"
+#include "logger.h"
+
+static size_t Number = 0;
+
+
+size_t GetBlockNumber()
```
Comment (mwoehlke-kitware) [inline src/BlockNumbering.cpp]: Especially in light of #1843, I wonder where we are planning to go here? At the least, it is a little unusual for functions to start with a capital letter; usually, code that uses CamelCase reserves names with an initial upper-case letter for classes, e.g. `ClassName` and `functionName`.

(And *mixing* CamelCaseWords and words_with_underscores is even more unusual...)

File: tests/cli/output/p.txt [hunk 2]
```diff
@@ -3,62 +3,62 @@ newlines                        = crlf
...
-#  13>            NEWLINE[               NONE][ 36/ 28/  1/  0][1/1/0][         2][1-0]
-#  14>               TYPE[               NONE][  9/  1/  9/  0][1/1/0][    470402][0-0]         unsigned
-#  14>               TYPE[               NONE][ 18/ 10/ 14/  1][1/1/0][    410402][0-0]                  long
-#  14>               WORD[               NONE][ 23/ 15/ 24/  1][1/1/0][   1800402][0-0]                       nTelIndex
-#  14>          SEMICOLON[               NONE][ 32/ 24/ 25/  0][1/1/0][ 100000402][0-0]                                ;
-#  14>            NEWLINE[               NONE][ 33/ 25/  1/  0][1/1/0][         2][1-0]
-#  15>        BRACE_CLOSE[             STRUCT][  1/  1/  2/  0][0/0/0][ 100000400][0-0] }
-#  15>          SEMICOLON[             STRUCT][  2/  2/  3/  0][0/0/0][ 100000000][0-0]  ;
-#  15>            NEWLINE[               NONE][  3/  3/  1/  0][0/0/0][         0][2-0]
+# Line                Tag              Parent          Columns Br/Lvl/pp     Flag   Nl  blockNu Text
```
Comment (mwoehlke-kitware) [inline tests/cli/output/p.txt]: I just noticed you are not capitalizing here, while the other column headers are all capitalized. This also makes me wonder if the "Nu" is really needed here; just "Block" might be better. (Note: this is prose and should not be subject to the same style rules as code.)

Off topic for this PR, but now that I'm thinking about it, "pp" being all lower case seems a little odd as well :wink:.

**Comment by CDanU** [comment]
I am somewhat concerned about the changes that are made in this PR. 
1. It adds state, on top of that also global state
2. it does not manage the state  
    - currently vbraces are ignored
    - _theoretically_ the numbering could change at some point but there are no blockNr updates (does that happen?)
3. Do we actually need numbering here or do we simply want to know if we are in the same block or not? 

Could one solution be to add chunks inside parenthesis into their own stack?

**Comment by guy-maurel** [comment]
Answers to @CDanU 
1. where is the problem?
2.a. the vbraces could be introduce, not yet necessary for the set of tests we have
2.b. the numbering is only a run time information. It is not stored. What is mean about "change" ?
3. Yes, I need it to solve the bug, successfully.
Any other solving solution is welcome.

**Comment by mwoehlke-kitware** [comment]
Hmm, this sounds like maybe I should jump in with some thoughts I've been having... one thing I've noticed about the parsing that seems to make things overly difficult is that there is not much in the way of tracking related tokens, or groups of tokens that form specific grammatical constructs.

W.r.t. #1843, I have been wondering if it would be possible to restructure the token list so that it is actually a tree, where each branch would represent a grammar. (For compatibility, I would implement separate methods to get the depth-first next token, i.e. essentially iterators over leaves, but it would also be possible for parsing and new formatting code to query the tree structure.)

For example, consider a current parsing result:

```
          Tag     Parent Br/Lvl/pp     Flag  Text
         WORD       NONE][0/0/0][     60000] Foo
      COMPARE       NONE][0/0/0][ 100000000]    <
         TYPE       NONE][0/0/0][     40000]     int
        COMMA       NONE][0/0/0][ 100000000]        ,
         WORD       NONE][0/0/0][     40000]          Bar
      COMPARE       NONE][0/0/0][ 100000000]             <
         TYPE       NONE][0/0/0][     40000]              int
        ARITH       NONE][0/0/0][ 100000000]                 >>
    FUNC_CALL       NONE][0/0/0][     40000]                    foo
  FPAREN_OPEN  FUNC_CALL][0/0/0][ 100000000]                       (
         WORD       NONE][0/1/0][     40010]                        Bar
      COMPARE       NONE][0/1/0][ 100000010]                           <
         TYPE       NONE][0/1/0][     40010]                            int
      COMPARE       NONE][0/1/0][ 100000010]                               >
        COMMA       NONE][0/1/0][ 100040010]                                ,
         TYPE       NONE][0/1/0][     40010]                                  int
 FPAREN_CLOSE  FUNC_CALL][0/0/0][ 100000000]                                     )
    SEMICOLON       NONE][0/0/0][ 100000000]                                      ;
      NEWLINE       NONE][0/0/0][         0]
```

I would *love* if this was instead represented like:

```
Tag                     Text
FUNC_PROTO              Foo<int, Bar<int>> foo(Bar<int>, int);
  FUNC_PROTO_RTYPE      Foo<int, Bar<int>>
    TTYPE               Foo<int, Bar<int>>
      TYPE              Foo
      ANGLE_OPEN           <
      TYPE                  int
      COMMA                    ,
      TTYPE                      Bar<int>
        TYPE                     Bar
        ANGLE_OPEN                  <
        TYPE                         int
        ANGLE_CLOSE                     >
      ANGLE_CLOSE                        >
  WORD                                     foo
  FUNC_PROTO_PARAMS                           (Bar<int>, int);
    FPAREN_OPEN                               (
    FUNC_PARAM                                 Bar<int>
      TTYPE                                    Bar<int>
        TYPE                                   Bar
        ANGLE_OPEN                                <
        TYPE                                       int
        ANGLE_CLOSE                                   >
    COMMA                                              ,
    FUNC_PARAM                                           int
      TYPE                                               int
    FPAREN_CLOSE                                            )
SEMICOLON                                                    ;
```

(I got sick of fiddling with that, but there should probably be a TTYPE_PARAMS in there, also.)

This would give us much better and consistent recognition of some things than we have now (for example, except at the stage that's trying to detect functions, I can't trivially ask if a token sequence is a function parameter, and the ability to skip complete grammar constructs is limited. Having a true tree would make it much easier to do things like find matching braces/parens/etc. or skip over "same level" grammar constructs. Compare, for example, the difficulty of parsing `foo<int, Bar<int>>` into "a template, whose proper name is `foo` and whose arguments are `int` and `Bar<int>` today, vs. with the above.

I'm not saying we should throw out this PR, but it's something to think about going forward...

**Comment by CDanU** [comment]
@gmaurel 
The problem is that while this will work for the test case that you had in mind but it opens up at least two new bugs. 

Look at `src/chunk_list.h` here you will see that both params aren't aligned anymore, this is at least worthy of discussion. Do we want the previous behaviour or not? 
I also would classify that as a sideeffect of the bug.

You can't just say that it is not necessary for the tests that we have (2a). In that case your tests are insufficient. 
I also don't get the comment about 2b well yes almost anything besides some few constexpr are hold during the runtime. That is not the point. You have introduced:
```cpp
static size_t number = 0;
```
into the code and while it is static it is still a global that is going to be a **nuisance** to debug. What I meant to say with 2b is that you currently are ignoring vbraces, those vbraces can be converted to normal braces. If that happens your numbering is off. Are you sure that this can't happen after you called the `numberTheBlocks` function? You can't really call it again this way because there is nothing in place to reset that global counter. This looks like something that need a class.

Did I understand you correctly about 3? You do need the actual numbering to solve this issue? It sure does not look like it seeing this check:
```cpp
if (pc->blockNumber != t2->blockNumber)
```

I will look into this later this night to see if I can find a better solution. 

@mwoehlke-kitware That would be nice, you should add that to the wiki or it gets lost here.

**Comment by guy-maurel** [comment]
Thanks to @CDanU for the comment.

As you proposed, I have a look at `src/chunk_list.h` but don't anderstand what is the problem you are speaking about. I run the script `scripts/Run_uncrustify_for_sources.sh` but don't get any error.
Right _"In that case your tests are insufficient."_
Many tests are insufficient and we get many new issues at the forum.

Just now we have the issue #1760 . The issue shows a problem,
the tests/input/cpp/templates.cpp shows also the same problem.

I think, the problem cannot be solve with the current "level" informations.
We need something else. I make a try with the blocks. The check
```.cpp
if (pc->blockNumber != t2->blockNumber)
```
applies on the two assigment sign to decide if we could align or not.

Let us know what is your proposal. I think a better one is always welcome.

**Comment by CDanU** [comment]
Ok, what do you think about this:
<code>

```diff
diff --git a/src/align.cpp b/src/align.cpp
index 8b6028b67..66c601c2b 100644
--- a/src/align.cpp
+++ b/src/align.cpp
@@ -792,7 +792,8 @@ chunk_t *align_assign(chunk_t *first, size_t span, size_t thresh, size_t *p_nl_c
       }
       else if (  equ_count == 0
               && chunk_is_token(pc, CT_ASSIGN)
-              && ((pc->flags & PCF_IN_TEMPLATE) == 0))  // Issue #999
+              && ((pc->flags & PCF_IN_TEMPLATE) == 0) // Issue #999
+              && pc->parent_type != CT_QUALIFIER)     // don't align '= 0' of a virtual function
       {
          LOG_FMT(LALASS, "%s(%d): orig_line is %zu, orig_col is %zu, pc->text() '%s'\n",
                  __func__, __LINE__, pc->orig_line, pc->orig_col, pc->text());
diff --git a/src/combine.cpp b/src/combine.cpp
index c6a562d5e..bbb778d37 100644
--- a/src/combine.cpp
+++ b/src/combine.cpp
@@ -1416,6 +1416,34 @@ void do_symbol_check(chunk_t *prev, chunk_t *pc, chunk_t *next)
 
    if (language_is_set(LANG_CPP))
    {
+      if(   chunk_is_token(pc, CT_ASSIGN)
+         && chunk_is_token(prev, CT_FPAREN_CLOSE)
+         && prev->parent_type == CT_FUNC_PROTO)
+      {
+          chunk_t * paren_open = chunk_skip_to_match_rev(prev);
+          const size_t po_level = paren_open != nullptr ? paren_open->level : 0;
+
+          bool found_virtual = false;
+          for( chunk_t* pprev = chunk_get_prev(paren_open)
+             ; (  pprev != nullptr
+               && pprev->level >= po_level
+               && pprev->type != CT_SEMICOLON
+               && pprev->type != CT_PRIVATE_COLON)
+             ; pprev = chunk_get_prev(pprev))
+          {
+              if(pprev->type == CT_QUALIFIER && chunk_is_str(pprev, "virtual", 7))
+              {
+                  found_virtual = true;
+                  break;
+              }
+          }
+
+          if(found_virtual)
+          {
+              pc->parent_type = CT_QUALIFIER;
+          }
+      }
+
       // Detect a braced-init-list
       if (  chunk_is_token(pc, CT_WORD)
          || chunk_is_token(pc, CT_TYPE)
```

</code>

This should be a somewhat cleaner solution for the problem at hand.

**Comment by mihaipopescu** [comment]
I see no difference

**Comment by guy-maurel** [comment]
@mihaipopescu 
The difference is: should the assignment sign = aligned or not aligned at these two lines?

**Comment by CDanU** [comment]
I think it could be aligned if we introduce a new option for that.
That alignment however should only be done in its own alignment stack that only ads chunks that are inside parenthesis that have the same (parenthesis?-)level.

**Comment by mwoehlke-kitware** [comment]
Um... FYI, I feel like an idiot that I only just realized this, but... `foo() = delete` is probably in the same boat. I think I'm leaning more and more to the approach I suggested in #1888, which is to just never align a `CT_ASSIGN` if the previous token is a `FPAREN_CLOSE:FUNC_PROTO`. (Not, anyway, with the same option as we have now.)

**Comment by arneforlie** [comment]
I can definitively confirm that.
```
class X19
{
public:
X19();
~X19() = default;
X19(const X19& rhs) = default;
X19& operator=(const X19& rhs) = default;
X19(X19&& rhs) = delete;
X19& operator=(X19&& rhs) = delete;
virtual void f(int x, int y) = 0;
void g(int x = 0);
};
```
is incorrectly formatted as:
```
class X19
{
public:
X19();
~X19()                         = default;
X19(const X19& rhs)            = default;
X19& operator=(const X19& rhs) = default;
X19(X19&& rhs)                 = delete;
X19& operator=(X19&& rhs)      = delete;
virtual void f(int x, int y)   = 0;
void g(int x                   = 0);
};
```

I will update #1760 to include this case too.

**Comment by CDanU** [comment]
Yes, I forgot about this bug ... I'll assign it to myself and will improve upon my above first attempt.

**Comment by gmaurel** [comment]
Replace by #1925

**Pull request closed without merging**